feat: replace stdlib poseidon implementation with optimized version

TomAFrench opened the pull request:
# Description

## Problem\*

Resolves <!-- Link to GitHub Issue -->

## Summary\*

This PR pulls out the existing poseidon implementation and replaces it with @zac-williamson's improved version.

I've removed the "optimized" version which hardcodes the values of `rf`/`rp` as it was equivalent to just using a constant config (something which I expect would happen in 99.9999% of cases).

## Additional Context



## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

vezenovm commented on noir_stdlib/src/hash/poseidon.nr:
```diff
@@ -3,66 +3,146 @@ use crate::field::modulus_num_bits;
+        for i in 0..state.len() {
+            state[i] = state[i] + C[state.len() * (_r + 1) as u64 + i as u64];
+        }
+        {
```
These feel like they can be removed but it is stylistic and non-blocking

vezenovm commented on noir_stdlib/src/hash/poseidon.nr:
```diff
@@ -3,66 +3,146 @@ use crate::field::modulus_num_bits;
-    ark: [Field; M], // Additive round keys
-    mds: [Field; N] // MDS Matrix in row-major order
+    C: [Field; N],
+    Mat: [[Field;M]; M], // Additive round keys
+    PMat: [[Field;M]; M],
+    S: [Field; X],
```
```suggestion
    Mat: [[Field; M]; M], // Additive round keys
    PMat: [[Field; M]; M],
    S: [Field; X],
```
Nit

Also it would be nice to have comments on `C`, `PMat`, and `S` as well

TomAFrench replied:
Have asked Zac to comment with some explanation/better naming for these fields.

vezenovm commented on noir_stdlib/src/hash/poseidon.nr:
```diff
@@ -3,66 +3,146 @@ use crate::field::modulus_num_bits;
+    Mat: [[Field; M]; M], // Additive round keys
+    PMat: [[Field; M]; M],
+    S: [Field; X]
+) -> PoseidonConfig<M,N,X> {
```
```suggestion
) -> PoseidonConfig<M, N, X> {
```
nit

vezenovm commented on noir_stdlib/src/hash/poseidon/bn254.nr:
```diff
@@ -2,93 +2,14 @@
-// Optimised permutation for this particular field; uses hardcoded rf and rp values,
-// which should agree with those in pos_conf.
-#[field(bn254)]
-pub fn permute<M, N, O>(pos_conf: PoseidonConfig<M, N>, mut state: [Field; O]) -> [Field; O] {
```
Do we not want a bn254 permute anymore?

vezenovm replied:
Oh wait did you remove this cause we actually require bn254 on the all the methods in `noir_stdlib/src/hash/poseidon/bn254/perm.nr`? Ignore this then

TomAFrench replied:
I removed this as this was just exactly the same as the generic permute but had hardcoded constants (something that we get for free if the config is a constant). I've then removed this and kept the generic version.

TomAFrench commented:
I've made a couple of updates to this PR:

- Added some documentation
- removed a generic type as the size of the MDS matrix is equal to the state vector
- renamed this generic to be `T` to make it clearer that it's the state vector length
- replaced usage of `state.len()` with `T`
- pulled out an `apply_matrix` helper function for applying the MDS matrices.
